distributed/tests/test_scheduler.py

```diff
@@ -872,6 +872,29 @@ def test_coerce_address():
     yield [w._close() for w in [a, b, c]]
 
 
+@gen_test()
+def test_workers_set():
+    s = Scheduler(validate=True)
+    s.start(0)
+    a = Worker(s.ip, s.port, name='alice', ip='127.0.0.1')
+    b = Worker(s.ip, s.port, name=123, ip='127.0.0.2')
+    c = Worker(s.ip, s.port, name='charlie', ip='127.0.0.2')
+    yield [a._start(), b._start(), c._start()]
+    # Manually setup environment state, just for this test
+    s.environments['test'] = b'not a real pickle string'
+    s.environment_workers['test'] = set([a.address, c.address])
```

Comment:
FWIW I used to do unit-testy things like this and eventually gave it up and just used clients all the time.  The internal state changes more quickly than the API and I found myself editing tests very often.  I'm not against this code by the way, it's great and arguably the "right" way to do things, just sharing my experience.  I now use clients all the time